src/main/java/org/opensearch/plugin/insights/rules/transport/top_queries/TransportTopQueriesAction.java

```diff
@@ -68,27 +73,122 @@ public TransportTopQueriesAction(
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                finalListener.onFailure(e);
+            }
+        };
+    }
+
+    protected void handleInMemoryDataResponse(
+        TopQueriesRequest request,
+        TopQueriesResponse inMemoryQueriesResponse,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        List<TopQueries> inMemoryTopQueries = inMemoryQueriesResponse.getNodes();
+        List<FailedNodeException> inMemoryDataFailures = inMemoryQueriesResponse.failures();
+        String from = request.getFrom();
+        String to = request.getTo();
+        if (from != null && to != null) {
+            fetchHistoricalData(request, inMemoryTopQueries, inMemoryDataFailures, finalListener);
+        } else {
+            finalListener.onResponse(inMemoryQueriesResponse);
+        }
+    }
+
+    protected void fetchHistoricalData(
+        TopQueriesRequest request,
+        List<TopQueries> inMemoryTopQueries,
+        List<FailedNodeException> inMemoryDataFailures,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        String from = request.getFrom();
+        String to = request.getTo();
+        String id = request.getId();
+        Boolean verbose = request.getVerbose();
+        final ThreadContext.StoredContext storedContext = threadPool.getThreadContext().stashContext();
+        try {
+            queryInsightsService.getTopQueriesService(request.getMetricType())
+                .getTopQueriesRecordsFromIndex(from, to, id, verbose, new ActionListener<List<SearchQueryRecord>>() {
+                    @Override
+                    public void onResponse(List<SearchQueryRecord> historicalRecords) {
+                        onHistoricalDataResponse(
+                            request,
+                            inMemoryTopQueries,
+                            inMemoryDataFailures,
+                            historicalRecords,
+                            finalListener,
+                            storedContext
+                        );
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        onHistoricalDataFailure(request, inMemoryTopQueries, inMemoryDataFailures, e, finalListener, storedContext);
+                    }
+                });
+        } catch (Exception e) {
+            storedContext.restore();
+            logger.error("Synchronous failure while initiating historical top queries fetch", e);
+            finalListener.onFailure(e);
+        }
+    }
+
+    protected void onHistoricalDataResponse(
+        TopQueriesRequest request,
+        List<TopQueries> inMemoryTopQueries,
+        List<FailedNodeException> inMemoryDataFailures,
+        List<SearchQueryRecord> historicalRecords,
+        ActionListener<TopQueriesResponse> finalListener,
+        ThreadContext.StoredContext storedContext
+    ) {
+        storedContext.restore();
```

Comment:
I think the `StoredContext` method param and the restore call on it can be avoided with above suggestion